lib/addons/argocd/index.ts

```diff
@@ -1,26 +1,359 @@
+            return "";
+        }
+
+        const appRepo = this.options.bootstrapRepo!;
+        let credentials = { url: btoa(appRepo.repoUrl) };
+
+        const secretValue = await this.getSecretValue(secretName, clusterInfo.cluster.stack.region);
+
+        let result = "";
+
+        switch (appRepo?.credentialsType) {
+            case "SSH":
+                credentials = { ...credentials, ...{ sshPrivateKey: btoa(secretValue) } };
+                result = sshRepoRef(appRepo.repoUrl, secretName);
+                break;
+            case "USERNAME":
+            case "TOKEN":
+                // eslint-disable-next-line no-case-declarations
+                const secretJson: any = JSON.parse(secretValue);
+                credentials = { ...credentials, ...{
+                        username: btoa(secretJson["username"]),
+                        password: btoa(secretJson["password"])
+                    }
+                };
+                result = userNameRepoRef(appRepo.repoUrl, secretName);
+                break;
+        }
+
+        const manifest = new KubernetesManifest(clusterInfo.cluster.stack, "argo-bootstrap-secret", {
+            cluster: clusterInfo.cluster,
+            manifest: [{
+                apiVersion: "v1",
+                kind: "Secret",
+                metadata: {
+                    name: secretName,
+                    namespace: this.options.namespace,
+                    labels: {
+                        "argocd.argoproj.io/secret-type": "repo-creds"
+                    }
+                },
+                data: credentials,
+            }],
+            overwrite: true,
+        });
+
+        manifest.node.addDependency(dependency);
+        return result;
+    }
+
+    /**
+     * Gets secret value from AWS Secret Manager. Requires access rights to the secret, specified by the secretName parameter.
+     * @param secretName name of the secret to retrieve
+     * @param region 
+     * @returns 
+     */
+    protected async getSecretValue(secretName: string, region: string): Promise<string> {
+        const secretManager = new SecretsManager({ region: region });
+        let secretString = "";
+        try {
+            let response = await secretManager.getSecretValue({ SecretId: secretName }).promise();
+            if (response) {
+                if (response.SecretString) {
+                    secretString = response.SecretString;
+                } else if (response.SecretBinary) {
+                    throw new Error(`Invalid secret format for ${secretName}. Expected string value, received binary.`);
+                }
+            }
+            return secretString;
+        }
+        catch (error) {
+            console.log(error);
+            throw error;
+        }
+    }
+
+    /**
+     * Creates a service account that can access secrets
+     * @param clusterInfo 
+     * @returns 
+     */
+    protected createServiceAccount(clusterInfo: ClusterInfo): ServiceAccount {
```

Comment:
seems like this is a utils function other addons would likely be able to use?

Reply:
This SA is specific to secret manager and it has a pretty broad permission set. Other service accounts that we generate (e.g. for Apoorva's CSI / secrets addon) are specific to a resource. But if we move out the policy out of this method, there is not much left compared to `cluster.addServiceAccount`.